File: scripts/run_ancestry_plink_gwas.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(data.table)
  library(ggplot2)
})

args <- commandArgs(trailingOnly = TRUE)

get_arg <- function(flag, default = NULL, required = FALSE) {
  idx <- which(args == flag)
  if (length(idx) == 0) {
    if (required) stop(sprintf("Missing required argument: %s", flag), call. = FALSE)
    return(default)
  }
  if (idx[length(idx)] == length(args)) {
    stop(sprintf("Missing value for argument: %s", flag), call. = FALSE)
  }
  args[idx[length(idx)] + 1]
}

parse_bool <- function(value) {
  tolower(trimws(as.character(value))) %in% c("1", "true", "yes", "y")
}

extract_base_iid <- function(sample_id) {
  x <- as.character(sample_id)
  us <- gregexpr("_", x, fixed = TRUE)[[1]]
  if (length(us) > 0 && us[1] != -1) {
    for (p in us) {
      left <- substr(x, 1, p - 1)
      right <- substr(x, p + 1, nchar(x))
      if (left == right) return(left)
    }
  }
  x
}

run_cmd <- function(cmd, cmd_args) {
  res <- system2(cmd, cmd_args, stdout = TRUE, stderr = TRUE)
  status <- attr(res, "status")
  if (!is.null(status) && status != 0) {
    stop(
      paste0(
        "Command failed: ", cmd, " ", paste(cmd_args, collapse = " "), "\n",
        paste(res, collapse = "\n")
      ),
      call. = FALSE
    )
  }
  invisible(res)
}

chr_sort_key <- function(ch) {
  s <- gsub("^chr", "", as.character(ch))
  if (grepl("^[0-9]+$", s)) return(c(0, as.integer(s)))
  c(1, match(s, unique(s)))
}

main <- function() {
  ancestry <- get_arg("--ancestry", required = TRUE)
  chromosomes <- strsplit(get_arg("--chromosomes", required = TRUE), ",", fixed = TRUE)[[1]]
  chromosomes <- chromosomes[nzchar(chromosomes)]
  input_dir <- get_arg("--input-dir", required = TRUE)
  output_dir <- get_arg("--output-dir", required = TRUE)
  merged_prefix <- get_arg("--merged-prefix", required = TRUE)
  gwas_prefix <- get_arg("--gwas-prefix", required = TRUE)
  gwas_tsv <- get_arg("--gwas-tsv", required = TRUE)
  manhattan_png <- get_arg("--manhattan-png", required = TRUE)
  phenotype <- get_arg("--phenotype", required = TRUE)
  phenotype_format <- get_arg("--phenotype-format", required = TRUE)
  table_iid_column <- get_arg("--table-iid-column", default = "IID")
  table_phenotype_column <- get_arg("--table-phenotype-column", default = "PHENO")
  covariates_file <- get_arg("--covariates-file", required = TRUE)
  covariates_format <- get_arg("--covariates-format", default = "plink_eigenvec")
  covariates_has_header <- parse_bool(get_arg("--covariates-has-header", default = "false"))
  covariates_fid_column <- get_arg("--covariates-fid-column", default = "FID")
  covariates_iid_column <- get_arg("--covariates-iid-column", default = "IID")
  n_pcs <- as.integer(get_arg("--n-pcs", default = "5"))

  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  pfile_prefixes <- character()
  for (chr_name in chromosomes) {
    prefix <- file.path(input_dir, sprintf("chr%s.dosage", chr_name))
    if (file.exists(paste0(prefix, ".pgen")) && file.exists(paste0(prefix, ".pvar")) && file.exists(paste0(prefix, ".psam"))) {
      pfile_prefixes <- c(pfile_prefixes, prefix)
    }
  }
  if (length(pfile_prefixes) == 0) {
    stop(sprintf("No chromosome PLINK dosage files found in %s", input_dir), call. = FALSE)
  }

  if (length(pfile_prefixes) == 1) {
    run_cmd("plink2", c("--pfile", pfile_prefixes[1], "--allow-extra-chr", "--make-pgen", "--out", merged_prefix))
  } else {
    merge_list <- file.path(output_dir, "merge_list.txt")
    merge_lines <- vapply(
      pfile_prefixes[-1],
      function(pref) paste0(pref, ".pgen ", pref, ".pvar ", pref, ".psam"),
      FUN.VALUE = character(1)
    )
    writeLines(merge_lines, merge_list)
    run_cmd(
      "plink2",
      c("--pfile", pfile_prefixes[1], "--pmerge-list", merge_list, "--allow-extra-chr", "--make-pgen", "--out", merged_prefix)
    )
  }

  merged_psam <- paste0(merged_prefix, ".psam")
  psam <- fread(merged_psam, data.table = TRUE)
  iid_col <- if ("IID" %in% names(psam)) "IID" else if ("#IID" %in% names(psam)) "#IID" else NA_character_
  if (is.na(iid_col)) stop(sprintf("Could not find IID column in %s", merged_psam), call. = FALSE)
  merged_iids <- as.character(psam[[iid_col]])

  sample_map <- setNames(merged_iids, vapply(merged_iids, extract_base_iid, FUN.VALUE = character(1)))

  if (tolower(phenotype_format) == "fam") {
    fam <- fread(phenotype, header = FALSE, data.table = TRUE)
    if (ncol(fam) < 6) stop(sprintf("FAM file must have at least 6 columns: %s", phenotype), call. = FALSE)
    pheno <- data.table(IID_BASE = as.character(fam[[2]]), PHENO = as.character(fam[[6]]))
  } else {
    ptab <- fread(phenotype, data.table = TRUE)
    if (!(table_iid_column %in% names(ptab)) || !(table_phenotype_column %in% names(ptab))) {
      stop("Phenotype table missing configured IID/phenotype columns", call. = FALSE)
    }
    pheno <- data.table(IID_BASE = as.character(ptab[[table_iid_column]]), PHENO = as.character(ptab[[table_phenotype_column]]))
  }
  pheno[, IID := sample_map[IID_BASE]]
  pheno <- pheno[!is.na(IID) & nzchar(PHENO)]
  pheno[, FID := IID]
  pheno <- unique(pheno[, .(FID, IID, PHENO)], by = "IID")
  if (nrow(pheno) == 0) stop("No phenotype rows matched merged PLINK sample IDs", call. = FALSE)
  pheno_file <- file.path(output_dir, "phenotype_for_plink.tsv")
  fwrite(pheno, pheno_file, sep = "\t")

  if (tolower(covariates_format) != "plink_eigenvec") {
    stop(sprintf("Unsupported covariates format: %s", covariates_format), call. = FALSE)
  }
  if (covariates_has_header) {
    cov <- fread(covariates_file, data.table = TRUE)
  } else {
    cov <- fread(covariates_file, header = FALSE, data.table = TRUE)
    if (ncol(cov) < 3) stop("Covariates file must have at least FID IID PC1", call. = FALSE)
    setnames(
      cov,
      old = names(cov),
      new = c(covariates_fid_column, covariates_iid_column, paste0("PC", seq_len(ncol(cov) - 2)))
    )
  }
  pc_cols <- names(cov)[grepl("^PC", names(cov), ignore.case = TRUE)]
  pc_cols <- head(pc_cols, n_pcs)
  if (length(pc_cols) < n_pcs) {
    stop(sprintf("Requested %d PCs; found %d", n_pcs, length(pc_cols)), call. = FALSE)
  }
  cov_out <- data.table(IID_BASE = as.character(cov[[covariates_iid_column]]))
  cov_out[, IID := sample_map[IID_BASE]]
  cov_out <- cov_out[!is.na(IID)]
  cov_out[, FID := IID]
  for (pc in pc_cols) { 
    cov_tmp <- cov[, c(covariates_iid_column, pc), with = FALSE] 
    names(cov_tmp) <- c("IID_BASE", pc) 
    cov_out <- merge(cov_out, cov_tmp, by = "IID_BASE", all.x = TRUE) 
  }
  cov_out <- unique(cov_out[, c("FID", "IID", pc_cols), with = FALSE], by = "IID")
  if (nrow(cov_out) == 0) stop("No covariate rows matched merged PLINK sample IDs", call. = FALSE)
  cov_file <- file.path(output_dir, "covariates_for_plink.tsv")
  fwrite(cov_out, cov_file, sep = "\t")

  run_cmd(
    "plink2",
    c(
      "--pfile", merged_prefix,
      "--allow-extra-chr",
      "--pheno", pheno_file,
      "--pheno-name", "PHENO",
      "--covar", cov_file,
      "--covar-name", paste(pc_cols, collapse = ","),
      "--glm", "hide-covar",
      "--out", gwas_prefix
    )
  )

  glm_candidates <- Sys.glob(paste0(gwas_prefix, ".PHENO.glm.*"))
  if (length(glm_candidates) == 0) {
    stop(sprintf("No PLINK GLM output found for prefix: %s", gwas_prefix), call. = FALSE)
  }
  logistic_hits <- glm_candidates[grepl("\\.logistic", glm_candidates)]
  glm_file <- if (length(logistic_hits) > 0) logistic_hits[1] else glm_candidates[1]

  g <- fread(glm_file, data.table = TRUE)
  if ("TEST" %in% names(g)) g <- g[TEST == "ADD"]
  chr_col <- if ("#CHROM" %in% names(g)) "#CHROM" else if ("CHROM" %in% names(g)) "CHROM" else NA_character_
  if (is.na(chr_col) || !("POS" %in% names(g)) || !("ID" %in% names(g)) || !("P" %in% names(g))) {
    stop(sprintf("Unexpected GLM columns in %s", glm_file), call. = FALSE)
  }

  g[, P := as.numeric(P)]
  g <- g[!is.na(P) & P > 0]
  g[, CHR := gsub("^chr", "", as.character(get(chr_col)))]
  g[, POS := as.numeric(POS)]
  g <- g[!is.na(POS)]

  keep <- c("CHR", "POS", "ID", "P")
  extras <- c("A1", "A1_FREQ", "BETA", "SE", "OR", "Z_STAT", "T_STAT", "ERRCODE")
  keep <- c(keep, extras[extras %in% names(g)])
  g_out <- g[, ..keep]
  fwrite(g_out, gwas_tsv, sep = "\t")

  if (nrow(g_out) == 0) {
    p <- ggplot() +
      annotate("text", x = 0.5, y = 0.5, label = "No GWAS points to plot", size = 5) +
      xlim(0, 1) + ylim(0, 1) +
      theme_void() +
      ggtitle(sprintf("%s ancestry dosage GWAS", ancestry))
    ggsave(manhattan_png, p, width = 14, height = 6, dpi = 180)
    return(invisible(NULL))
  }

  chr_levels <- unique(g_out$CHR)
  chr_levels <- chr_levels[order(vapply(chr_levels, function(ch) chr_sort_key(ch)[1], numeric(1)),
                                 vapply(chr_levels, function(ch) chr_sort_key(ch)[2], numeric(1)))]

  offsets <- data.table(CHR = chr_levels)
  offsets[, chr_max := g_out[CHR == .BY$CHR, max(POS, na.rm = TRUE)], by = CHR]
  offsets[, offset := shift(cumsum(chr_max + 1e6), fill = 0)]
  g_plot <- merge(g_out, offsets[, .(CHR, offset, chr_max)], by = "CHR", all.x = TRUE)
  g_plot[, x := POS + offset]
  g_plot[, y := -log10(P)]
  g_plot[, chr_idx := as.integer(factor(CHR, levels = chr_levels))]

  centers <- offsets[, .(CHR, center = offset + chr_max / 2)]

  p <- ggplot(g_plot, aes(x = x, y = y, color = factor(chr_idx %% 2))) +
    geom_point(size = 0.6, alpha = 0.8) +
    geom_hline(yintercept = -log10(5e-8), linetype = "dashed", color = "#cc2f2f", linewidth = 0.4) +
    scale_color_manual(values = c("#1f77b4", "#ff7f0e"), guide = "none") +
    scale_x_continuous(breaks = centers$center, labels = centers$CHR) +
    labs(
      title = sprintf("%s ancestry dosage GWAS", ancestry),
      x = "Chromosome",
      y = "-log10(P)"
    ) +
    theme_bw(base_size = 12)

  ggsave(manhattan_png, p, width = 14, height = 6, dpi = 180)
}

main()

File: scripts/run_ancestry_plink_gwas.py
#!/usr/bin/env python3
import argparse
import glob
import os
import subprocess

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd


def parse_bool(value):
    return str(value).strip().lower() in {"1", "true", "yes", "y"}


def extract_base_iid(sample_id):
    x = str(sample_id)
    underscores = [i for i, ch in enumerate(x) if ch == "_"]
    for idx in underscores:
        left = x[:idx]
        right = x[idx + 1 :]
        if left == right:
            return left
    return x


def run_cmd(cmd, cwd=None):
    result = subprocess.run(cmd, cwd=cwd, check=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    if result.returncode != 0:
        raise RuntimeError(
            f"Command failed ({result.returncode}): {' '.join(cmd)}\n"
            f"STDOUT:\n{result.stdout}\nSTDERR:\n{result.stderr}"
        )
    return result


def chr_sort_key(ch):
    s = str(ch).replace("chr", "")
    return (0, int(s)) if s.isdigit() else (1, s)


def load_psam_iids(psam_path):
    psam = pd.read_csv(psam_path, sep="\t", dtype=str)
    iid_col = "IID" if "IID" in psam.columns else "#IID"
    if iid_col not in psam.columns:
        raise ValueError(f"Could not find IID column in {psam_path}")
    return psam[iid_col].astype(str).tolist()


def build_sample_map(sample_ids):
    mapping = {}
    for sid in sample_ids:
        mapping[extract_base_iid(sid)] = sid
    return mapping


def prepare_phenotype(args, sample_map, out_path):
    if args.phenotype_format == "fam":
        fam = pd.read_csv(args.phenotype, sep=r"\s+", header=None, dtype=str, engine="python")
        if fam.shape[1] < 6:
            raise ValueError(f"FAM must have >=6 columns: {args.phenotype}")
        raw = pd.DataFrame({"IID_BASE": fam.iloc[:, 1].astype(str), "PHENO": fam.iloc[:, 5].astype(str)})
    else:
        tab = pd.read_csv(args.phenotype, sep=None, dtype=str, engine="python")
        if args.table_iid_column not in tab.columns or args.table_phenotype_column not in tab.columns:
            raise ValueError("Phenotype table missing IID/PHENO configured columns")
        raw = pd.DataFrame(
            {
                "IID_BASE": tab[args.table_iid_column].astype(str),
                "PHENO": tab[args.table_phenotype_column].astype(str),
            }
        )

    raw["IID"] = raw["IID_BASE"].map(sample_map)
    pheno = raw.dropna(subset=["IID", "PHENO"]).copy()
    pheno["FID"] = pheno["IID"]
    pheno = pheno[["FID", "IID", "PHENO"]].drop_duplicates(subset=["IID"])
    pheno.to_csv(out_path, sep="\t", index=False)
    if pheno.empty:
        raise ValueError("No phenotype rows matched merged PLINK sample IDs")


def prepare_covariates(args, sample_map, out_path):
    if args.covariates_format != "plink_eigenvec":
        raise ValueError(f"Unsupported covariates format: {args.covariates_format}")

    has_header = parse_bool(args.covariates_has_header)
    if has_header:
        cov = pd.read_csv(args.covariates_file, sep=r"\s+", dtype=str, engine="python")
    else:
        cov = pd.read_csv(args.covariates_file, sep=r"\s+", header=None, dtype=str, engine="python")
        if cov.shape[1] < 3:
            raise ValueError("Covariates file needs at least FID IID PC1")
        cov.columns = [args.covariates_fid_column, args.covariates_iid_column] + [
            f"PC{i}" for i in range(1, cov.shape[1] - 1)
        ]

    pc_cols = [c for c in cov.columns if str(c).upper().startswith("PC")][: args.n_pcs]
    if len(pc_cols) < args.n_pcs:
        raise ValueError(f"Requested {args.n_pcs} PCs; found {len(pc_cols)}")

    out = pd.DataFrame()
    out["IID_BASE"] = cov[args.covariates_iid_column].astype(str)
    out["IID"] = out["IID_BASE"].map(sample_map)
    out = out.dropna(subset=["IID"]).copy()
    out["FID"] = out["IID"]
    for col in pc_cols:
        out[col] = pd.to_numeric(cov.loc[out.index, col], errors="coerce")

    out = out[["FID", "IID"] + pc_cols].drop_duplicates(subset=["IID"])
    out.to_csv(out_path, sep="\t", index=False)
    if out.empty:
        raise ValueError("No covariate rows matched merged PLINK sample IDs")
    return pc_cols


def choose_glm_file(prefix):
    candidates = sorted(glob.glob(prefix + ".PHENO1.glm.*"))
    if not candidates:
        raise FileNotFoundError(f"No PLINK GLM output found for prefix: {prefix}")
    for path in candidates:
        if ".logistic" in path:
            return path
    return candidates[0]


def make_gwas_table(glm_path, out_tsv):
    g = pd.read_csv(glm_path, sep=r"\s+", dtype=str, engine="python")
    if "TEST" in g.columns:
        g = g[g["TEST"] == "ADD"].copy()

    chr_col = "#CHROM" if "#CHROM" in g.columns else ("CHROM" if "CHROM" in g.columns else None)
    if chr_col is None or "POS" not in g.columns or "ID" not in g.columns or "P" not in g.columns:
        raise ValueError(f"Unexpected GLM columns in {glm_path}")

    g["P"] = pd.to_numeric(g["P"], errors="coerce")
    g = g[g["P"].notna() & (g["P"] > 0)].copy()
    g["CHR"] = g[chr_col].astype(str).str.replace("chr", "", regex=False)
    g["POS"] = pd.to_numeric(g["POS"], errors="coerce")
    g = g[g["POS"].notna()].copy()

    keep = ["CHR", "POS", "ID", "P"]
    for extra in ["A1", "A1_FREQ", "BETA", "SE", "OR", "Z_STAT", "T_STAT", "ERRCODE"]:
        if extra in g.columns:
            keep.append(extra)

    out = g[keep].copy()
    out.to_csv(out_tsv, sep="\t", index=False)
    return out


def plot_manhattan(gwas_df, png_path, title):
    if gwas_df.empty:
        plt.figure(figsize=(12, 5))
        plt.text(0.5, 0.5, "No GWAS points to plot", ha="center", va="center")
        plt.axis("off")
        plt.savefig(png_path, dpi=180, bbox_inches="tight")
        plt.close()
        return

    df = gwas_df.copy()
    df["CHR"] = df["CHR"].astype(str)
    df["POS"] = pd.to_numeric(df["POS"], errors="coerce")
    df["P"] = pd.to_numeric(df["P"], errors="coerce")
    df = df.dropna(subset=["POS", "P"])
    df = df[df["P"] > 0]
    if df.empty:
        plt.figure(figsize=(12, 5))
        plt.text(0.5, 0.5, "No valid GWAS p-values to plot", ha="center", va="center")
        plt.axis("off")
        plt.savefig(png_path, dpi=180, bbox_inches="tight")
        plt.close()
        return

    chr_order = sorted(df["CHR"].unique(), key=chr_sort_key)
    offsets = {}
    running = 0
    centers = []
    for ch in chr_order:
        ch_max = df.loc[df["CHR"] == ch, "POS"].max()
        offsets[ch] = running
        centers.append((ch, running + ch_max / 2.0))
        running += ch_max + 1_000_000

    df["x"] = df.apply(lambda r: r["POS"] + offsets[r["CHR"]], axis=1)
    df["y"] = -np.log10(df["P"].astype(float))

    plt.figure(figsize=(14, 6))
    colors = ["#1f77b4", "#ff7f0e"]
    for i, ch in enumerate(chr_order):
        sub = df[df["CHR"] == ch]
        plt.scatter(sub["x"], sub["y"], s=8, c=colors[i % 2], alpha=0.75, linewidths=0)

    plt.axhline(-np.log10(5e-8), color="#cc2f2f", linestyle="--", linewidth=1)
    plt.title(title)
    plt.xlabel("Chromosome")
    plt.ylabel("-log10(P)")
    plt.xticks([c for _, c in centers], [ch for ch, _ in centers], rotation=0)
    plt.tight_layout()
    plt.savefig(png_path, dpi=180)
    plt.close()


def main():
    parser = argparse.ArgumentParser(description="Run ancestry-specific merged PLINK GWAS and Manhattan plot")
    parser.add_argument("--ancestry", required=True)
    parser.add_argument("--chromosomes", required=True)
    parser.add_argument("--input-dir", required=True)
    parser.add_argument("--output-dir", required=True)
    parser.add_argument("--merged-prefix", required=True)
    parser.add_argument("--gwas-prefix", required=True)
    parser.add_argument("--gwas-tsv", required=True)
    parser.add_argument("--manhattan-png", required=True)
    parser.add_argument("--phenotype", required=True)
    parser.add_argument("--phenotype-format", choices=["fam", "table"], required=True)
    parser.add_argument("--fam-case-code", default="2")
    parser.add_argument("--fam-control-code", default="1")
    parser.add_argument("--table-iid-column", default="IID")
    parser.add_argument("--table-phenotype-column", default="PHENO")
    parser.add_argument("--covariates-file", required=True)
    parser.add_argument("--covariates-format", default="plink_eigenvec")
    parser.add_argument("--covariates-has-header", default="false")
    parser.add_argument("--covariates-fid-column", default="FID")
    parser.add_argument("--covariates-iid-column", default="IID")
    parser.add_argument("--n-pcs", type=int, default=5)
    args = parser.parse_args()

    os.makedirs(args.output_dir, exist_ok=True)
    chr_list = [c for c in str(args.chromosomes).split(",") if c]

    pfile_triples = []
    for ch in chr_list:
        prefix = os.path.join(args.input_dir, f"chr{ch}.dosage")
        pgen = prefix + ".pgen"
        pvar = prefix + ".pvar"
        psam = prefix + ".psam"
        if os.path.exists(pgen) and os.path.exists(pvar) and os.path.exists(psam):
            pfile_triples.append((pgen, pvar, psam, prefix))
    if not pfile_triples:
        raise FileNotFoundError(f"No chromosome PLINK dosage files found in {args.input_dir}")

    first_prefix = pfile_triples[0][3]
    if len(pfile_triples) == 1:
        run_cmd(
            [
                "plink2",
                "--pfile",
                first_prefix,
                "--allow-extra-chr",
                "--make-pgen",
                "--out",
                args.merged_prefix,
            ]
        )
    else:
        merge_list = os.path.join(args.output_dir, "merge_list.txt")
        with open(merge_list, "w", encoding="utf-8") as handle:
            for pgen, pvar, psam, _ in pfile_triples[1:]:
                handle.write(f"{pgen} {pvar} {psam}\n")

        run_cmd(
            [
                "plink2",
                "--pfile",
                first_prefix,
                "--pmerge-list",
                merge_list,
                "--allow-extra-chr",
                "--make-pgen",
                "--out",
                args.merged_prefix,
            ]
        )

    merged_psam = args.merged_prefix + ".psam"
    sample_ids = load_psam_iids(merged_psam)
    sample_map = build_sample_map(sample_ids)

    pheno_file = os.path.join(args.output_dir, "phenotype_for_plink.tsv")
    covar_file = os.path.join(args.output_dir, "covariates_for_plink.tsv")
    prepare_phenotype(args, sample_map, pheno_file)
    pc_cols = prepare_covariates(args, sample_map, covar_file)

    run_cmd(
        [
            "plink2",
            "--pfile",
            args.merged_prefix,
            "--allow-extra-chr",
            "--pheno",
            pheno_file,
            "--pheno-name",
            "PHENO",
            "--covar",
            covar_file,
            "--covar-name",
            ",".join(pc_cols),
            "--glm",
            "hide-covar",
            "--out",
            args.gwas_prefix,
        ]
    )

    glm_file = choose_glm_file(args.gwas_prefix)
    gwas_df = make_gwas_table(glm_file, args.gwas_tsv)
    plot_manhattan(gwas_df, args.manhattan_png, f"{args.ancestry} ancestry dosage GWAS")


if __name__ == "__main__":
    main()
